packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+    if (_rawValues.isEmpty) {
+      _rawData.remove(_valuesMapKey);
+    }
+    if (needsUpdate) {
+      _manager?.scheduleSerialization();
+    }
+    return result;
+  }
+
+  /// Checks whether a value stored in the bucket under the provided `id`.
+  ///
+  /// See also:
+  ///
+  ///  * [get], which retrieves a stored value from the bucket.
+  ///  * [put], which stores a value in the bucket.
+  ///  * [remove], which removes a value from the bucket.
+  bool containsValue(RestorationId id) {
+    assert(id != null);
+    return _rawValues.containsKey(id.value);
+  }
+
+  // Child management.
+
+  // The restoration IDs and associated buckets of children that have been
+  // claimed via [claimChild].
+  final Map<RestorationId, RestorationBucket> _claimedChildren = <RestorationId, RestorationBucket>{};
+  // Newly created child buckets that are scheduled to be added to [_rawChildren] in
+  // [_finalize], see comment in [claimChild] for details.
+  final Map<RestorationId, Set<RestorationBucket>> _childrenToAdd = <RestorationId, Set<RestorationBucket>>{};
+
+  /// Claims ownership of the child with the provided `id` from this bucket.
+  ///
+  /// If the application is getting restored to a previous state, the bucket
+  /// will contain all the data that was previously stored in the bucket.
+  /// Otherwise, an empty bucket is returned.
+  ///
+  /// The claimer of the bucket is expected to use the data stored in the bucket
+  /// to restore itself to its previous state described by the data in the
+  /// bucket. If the bucket is empty, it should initialize itself to default
+  /// values. Whenever the information, that the claimer needs to restore its
+  /// state changes, the data in the bucket should be updated to reflect that.
+  ///
+  /// A child bucket with a given `id` can only have one owner. If another owner
+  /// claims a child bucket with the same `id` an exception will be thrown at
+  /// the end of the current frame unless the previous owner has either deleted
+  /// its bucket by calling [dispose] or has moved it to a new parent via
+  /// [adoptChild].
+  ///
+  /// When the returned bucket is no longer needed, it must be [dispose]d to
+  /// delete the information stored in it from the app's restoration data.
+  RestorationBucket claimChild(RestorationId id, {@required Object debugOwner}) {
+    assert(id != null);
+    // There are three cases to consider:
+    // 1. Claiming an id that has already been claimed.
+    // 2. Claiming an id that doesn't yet exist in [_rawChildren].
+    // 3. Claiming an id that does exist in [_rawChildren] and hasn't been
+    //    claimed yet.
+    // If an id has already been claimed (case 1) the current owner may give up
+    // that id later this frame and it can be re-used. In anticipation of the
+    // previous owner's surrender of the id, we return an empty bucket for this
+    // new claim and check in [_finalize] that at the end of the frame the old
```

Comment:
So maybe a slightly different question than my one above in the explanation about finalizers - what if instead we just asserted that there was only one owner at the end of it all?